packages/material-ui-lab/src/DateRangePickerDay/DateRangePickerDay.tsx

```diff
@@ -43,96 +149,98 @@ const startBorderStyle = {
     ...startBorderStyle,
     paddingLeft: 0,
     marginLeft: DAY_MARGIN / 2,
-  },
-  rangeIntervalDayHighlightEnd: {
+  }),
+  ...((styleProps.isEndOfHighlighting || styleProps.isEndOfMonth) && {
     ...endBorderStyle,
     paddingRight: 0,
     marginRight: DAY_MARGIN / 2,
+  }),
+}));
+
+const DateRangePickerDayRangeIntervalPreview = styled(
+  'div',
+  {},
+  { name: 'MuiDateRangePickerDay', slot: 'RangeIntervalPreview' },
+)<{ styleProps: StyleProps }>(({ theme, styleProps }) => ({
+  // replace default day component margin with transparent border to avoid jumping on preview
+  border: '2px solid transparent',
+  ...(styleProps.isPreviewing &&
+    !styleProps.outsideCurrentMonth && {
+      borderRadius: 0,
+      border: `2px dashed ${theme.palette.divider}`,
+      borderLeftColor: 'transparent',
+      borderRightColor: 'transparent',
+      [`&.${dateRangePickerDayClasses.rangeIntervalDayPreviewStart}`]: {
+        borderLeftColor: theme.palette.divider,
+        ...startBorderStyle,
+      },
+      [`&.${dateRangePickerDayClasses.rangeIntervalDayPreviewEnd}`]: {
+        borderRightColor: theme.palette.divider,
+        ...endBorderStyle,
+      },
+    }),
+}));
+
+const DateRangePickerDayDay = (styled(
+  PickersDay,
+  {},
+  { name: 'MuiDateRangePickerDay', slot: 'Day' },
+)<{ styleProps: StyleProps }>(({ theme, styleProps }) => ({
+  // Required to overlap preview border
+  transform: 'scale(1.1)',
+  '& > *': {
+    transform: 'scale(0.9)',
   },
-  day: {
-    // Required to overlap preview border
-    transform: 'scale(1.1)',
-    '& > *': {
-      transform: 'scale(0.9)',
-    },
-  },
-  dayOutsideRangeInterval: {
+  ...(!styleProps.selected && {
+    backgroundColor: 'transparent',
+  }),
+  ...(!styleProps.isHighlighting && {
     '&:hover': {
       border: `1px solid ${theme.palette.grey[500]}`,
     },
-  },
-  dayInsideRangeInterval: {
-    color: theme.palette.getContrastText(alpha(theme.palette.primary.light, 0.6)),
-  },
-  notSelectedDate: {
-    backgroundColor: 'transparent',
-  },
-  rangeIntervalPreview: {
-    // replace default day component margin with transparent border to avoid jumping on preview
-    border: '2px solid transparent',
-  },
-  rangeIntervalDayPreview: {
-    borderRadius: 0,
-    border: `2px dashed ${theme.palette.divider}`,
-    borderLeftColor: 'transparent',
-    borderRightColor: 'transparent',
-    '&$rangeIntervalDayPreviewStart': {
-      borderLeftColor: theme.palette.divider,
-      ...startBorderStyle,
-    },
-    '&$rangeIntervalDayPreviewEnd': {
-      borderRightColor: theme.palette.divider,
-      ...endBorderStyle,
-    },
-  },
-  rangeIntervalDayPreviewStart: {},
-  rangeIntervalDayPreviewEnd: {},
-});
+  }),
+  ...(!styleProps.selected &&
+    styleProps.isHighlighting && {
+      color: theme.palette.getContrastText(alpha(theme.palette.primary.light, 0.6)),
+    }),
+})) as unknown) as <TDate>(
```

Comment:
`unknown` is needed for workaround with `styleProps`, otherwise got this error.
![image](https://user-images.githubusercontent.com/18292247/118769463-64cc6500-b8aa-11eb-8303-4b640e959240.png)

> I think to improve this, `styleProps` in `experimental_styled` should be a generic with default type of `Record<string, unknown>`